exporter/otlpexporter/internal/arrow/exporter.go

```diff
@@ -0,0 +1,216 @@
+		ready:       newStreamPrioritizer(settings),
+		returning:   make(chan *Stream, settings.NumStreams),
+		cancel:      func() {},
+	}
+}
+
+// Start creates the background context used by all streams and starts
+// a stream controller, which initializes the initial set of streams.
+func (e *Exporter) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+
+	e.cancel = cancel
+	e.wg.Add(1)
+	go e.runStreamController(ctx)
+
+	return nil
+}
+
+// runStreamController starts the initial set of streams, then waits for streams to
+// terminate one at a time and restarts them.  If streams come back with a nil
+// client (meaning that OTLP+Arrow was not supported by the endpoint), it will
+// not be restarted.
+func (e *Exporter) runStreamController(bgctx context.Context) {
+	defer e.cancel()
+	defer e.wg.Done()
+
+	running := e.settings.NumStreams
+
+	// Start the initial number of streams
+	for i := 0; i < running; i++ {
+		e.wg.Add(1)
+		go e.runArrowStream(bgctx)
+	}
+
+	for {
+		select {
+		case stream := <-e.returning:
+			if stream.client != nil {
+				// The stream closed or broken.  Restart it.
+				e.wg.Add(1)
+				go e.runArrowStream(bgctx)
+				continue
+			}
+			// Otherwise, the stream never got started.  It was
+			// downgraded and senders will use the standard OTLP path.
+			running--
+
+			// None of the streams were able to connect to
+			// an Arrow endpoint.
+			if running == 0 {
+				e.telemetry.Logger.Info("failed to establish OTLP+Arrow streaming, downgrading")
+				e.ready.downgrade()
+			}
+
+		case <-bgctx.Done():
+			// We are shutting down.
+			return
+		}
+	}
+}
+
+// runArrowStream begins one gRPC stream using a child of the background context.
+// If the stream connection is successful, this goroutine starts another goroutine
+// to call writeStream() and performs readStream() itself.  When the stream shuts
+// down this call synchronously waits for and unblocks the consumers.
+func (e *Exporter) runArrowStream(bgctx context.Context) {
+	ctx, cancel := context.WithCancel(bgctx)
+
+	stream := &Stream{
+		toWrite:  make(chan writeItem, 1),
+		producer: arrowRecord.NewProducer(),
+		waiters:  map[string]chan error{},
+		cancel:   cancel,
+	}
+
+	defer func() {
+		e.wg.Done()
+		cancel()
+		e.returning <- stream
+	}()
+
+	sc, err := e.client.ArrowStream(ctx, e.grpcOptions...)
+	if err != nil {
+		// TODO: only when this is a permanent (e.g., "no such
+		// method") error, downgrade to standard OTLP.
+		// Returning with stream.client == nil signals the
+		// lack of an Arrow stream endpoint.  When all the
+		// streams return with .client == nil, the ready
+		// channel will be closed.
+		e.telemetry.Logger.Error("cannot start event stream", zap.Error(err))
+		return
+	}
+	// Setting .client != nil indicates that the endpoint was valid,
+	// streaming may start.  When this stream finishes, it will be
+	// restarted.
+	stream.client = sc
+
+	var ww sync.WaitGroup
+
+	ww.Add(1)
+	e.wg.Add(1)
```

Comment:
it seems we never mark Done for the wg.Add here, should we be doing that in the write?

Reply:
Good catch. Will make sure to add test coverage, this wouldn't been easy to spot because shutdown would hang.

I think it's time to create an abstraction around the use of context cancelation and wait groups. Note that the stream doesn't have a pointer to the Exporter to Done() its waitgroup, so ... for now this reads

```
	ww.Add(1)
	e.wg.Add(1)
	go func() {
		defer e.wg.Done()
		stream.write(ctx, e, &ww)
	}()

```